src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Reload/ReloadableProject.cs

```diff
@@ -0,0 +1,144 @@
+                    // we want to let the normal solution reload occur so that the user sees the failure, and 2) it allows us to clear the current project
+                    // and do a deep copy from the newly loaded one to the original - replacing its contents. Note that the Open call is cached so that if
+                    // the project is already in a collection, it will just return the cached one. For this reason, and the fact we don't want the project to
+                    // appear in the global project collection, the file is opened in a new collection which we will discard when done.
+                    try
+                    {
+                        ProjectCollection thisCollection = new ProjectCollection();
+                        var newProject = ProjectRootElement.Open(_projectVsServices.Project.FullPath, thisCollection);
+
+                        msbuildProject.RemoveAllChildren();
+                        msbuildProject.DeepCopyFrom(newProject);
+
+                        // There isn't a way to clear the dirty flag on the project xml, so to work around that the project is saved
+                        // to a StringWriter. 
+                        var tw = new StringWriter();
+                        msbuildProject.Save(tw);
+
+                        thisCollection.UnloadAllProjects();
+                    }
+                    catch (Microsoft.Build.Exceptions.InvalidProjectFileException)
+                    {
+                        // Indicate we weren't able to complete the action. We want to do a normal reload 
+                        return ProjectReloadResult.ReloadFailed;
+                    }
+                    catch (Exception ex)
+                    {
+                        // Any other exception likely mean the msbuildProject is not in a good state. Example, DeepCopyFrom failed.
+                        // The only safe thing to do at this point is to reload the project in the solution
+                        // TODO: should we have an additional return value here to indicate that the existing project could be in a bad
+                        // state and the reload needs to happen without the user being able to block it?
+                        System.Diagnostics.Debug.Assert(false, "Replace xml failed with: " + ex.Message);
+                        return ProjectReloadResult.ReloadFailed;
+                    }
+                }
+            }
+
+            // It is important to wait for the new tree to be published to ensure all updates have occurred before the reload manager
+            // releases its hold on the UI thread. This prevents unnecessary race conditions and prevent the user
+            // from interacting with the project until the evaluation has completed.
+            await _projectVsServices.ProjectTree.TreeService.PublishLatestTreeAsync(blockDuringLoadingTree: true).ConfigureAwait(false);
```

Comment:
It seems to be fine. It just means we will block here until the tree is published. This was a recommendation from Lifeng.  I noticed the same technique used elsewhere in CPS. Like everything else we will need to  measure this.